src/irmin-pack/ext.ml

```diff
@@ -338,6 +338,20 @@ struct
     in
     Checks.integrity_check ?ppf ~auto_repair ~check t.index
 
+  let integrity_check_inodes ?ppf t :
+      (unit, [> `Cannot_fix of string | `Corrupted of int ]) result =
+    let module Checks = Store.Checks (Index) in
+    let nodes = X.Repo.node_t t |> snd in
+    let check ~kind ~offset ~length k =
+      match kind with
+      | `Contents -> Ok ()
+      | `Node -> X.Node.CA.integrity_check_inodes ~offset ~length k nodes
```

Comment:
If i'm not mistaken, this is the source of the problem for your slow check on tezos.

`X.Node.CA.integrity_check_inodes` is called for all objects in the pack file with the magic `'I'` or `'N'`, but `X.Node.CA.integrity_check_inodes` itself recursively loads all the tree below.

For each inode tree made of N objects, the number of calls to `Pack.unsafe_find` is not O(N), but O(N*N)

Reply:
As discussed offline, I modified this and we do a graph traversal instead to ensure that we check only the "root" nodes for inodes instead of checking all the intermediate nodes. 

I'm running it on the 42G store and after around 3 hours I'm at 
`1252049k nodes / 3k commits` 
maybe I missed something again, and we can be faster? (the integrity check is also very slow, around 10 hours for 220k commits).